python/hail/methods/statgen.py

```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
+
+    5. For each row and column with a non-missing genotype call, :math:`E`, the
+       expected number of homozygotes (from population AAF), is computed as
+       :math:`1.0 - (2.0*maf*(1.0-maf))`.
+
+    6. For each row and column with a non-missing genotype call, :math:`O`, the
+       observed number of homozygotes, is computed interpreting ``0`` as
+       heterozygote and ``1`` as homozygote`
+
+    7. For each row and column with a non-missing genotype call, :math:`N` is
+       incremented by 1
+
+    8. For each column, :math:`E`, :math:`O`, and :math:`N` are combined across
+       variants
+
+    9. For each column, :math:`F` is calculated by :math:`(O - E) / (N - E)`
+
+    10. A sex is assigned to each sample with the following criteria:
+        - Female when ``F < 0.2``
+        - Male when ``F > 0.8``
+        Use `female_threshold` and `male_threshold` to change this behavior.
+
+    **Annotations**
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`.
+    - **is_female** (:class:`.TBoolean`) -- True if the imputed sex is female,
+      false if male, missing if undetermined.
+    - **f_stat** (:class:`.TFloat64`) -- Inbreeding coefficient.
+    - **n_called**  (:class:`.TInt64`) -- Number of variants with a genotype call.
+    - **expected_homs** (:class:`.TFloat64`) -- Expected number of homozygotes.
+    - **observed_homs** (:class:`.TInt64`) -- Observed number of homozygotes.
+
+    call : :class:`.CallExpression`
+        A genotype call for each row and column. The source dataset's row keys
+        must be [[locus], alleles] with types :class:`.TLocus` and
+        :class:`.ArrayStringExpression`. Moreover, the keys must all be biallelic.
+    aaf_threshold : :obj:`float`
+        Minimum alternate allele frequency threshold.
+    include_par : :obj:`bool`
+        Include pseudoautosomal regions.
+    female_threshold : :obj:`float`
+        Samples are called females if F < female_threshold.
+    male_threshold : :obj:`float`
+        Samples are called males if F > male_threshold.
+    aaf : :obj:`str` or :obj:`None`
+        A field defining the alternate allele frequency for each row. If
+        ``None``, AAF will be computed from `call`.
+
+    Return
+    ------
+    :class:`.Table`
+        Sex imputation statistics per sample.
+
+    """
+
+    f = functions
+
+    ds = call._indices.source
+    ds, _ = ds._process_joins(call)
+    ds = ds.annotate_entries(call = call)
+    ds = require_biallelic(ds, 'impute_sex')
+    ds = require_variant(ds, 'impute_sex')
+    def locus(ds):
+        # FIXME use keys
+        return ds.v.locus() # ds[ds.row_key[0]]
+    gr = locus(ds).dtype.reference_genome
+
+    if (aaf is None):
+        ds = ds.annotate_rows(
+            aaf=(agg.sum(ds.call.num_alt_alleles()).to_float64() /
+                 agg.count_where(f.is_defined(ds.call)) / 2))
```

Comment:
should this be `* 2` rather than `/ 2`?

Reply:
divison is left associative so no, but I rewrote it to
```
(... /
 (2.0 * ...))
```
so hopefully that's more clear

Reply:
can we use call stats?